imap_processing/ultra/l0/decom_ultra.py

```diff
@@ -1,269 +1,226 @@
-
-    Parameters
-    ----------
-    decom_data : dict
-        Dictionary to which the data is appended.
-    packet : space_packet_parser.packets.CCSDSPacket
-        Individual packet.
-    """
-    for key, value in packet.user_data.items():
-        decom_data[key].append(value)
-
-    ccsds_data = CcsdsData(packet.header)
-    append_ccsds_fields(decom_data, ccsds_data)
-
-
-def process_ultra_apids(data: list, apid: int) -> Union[dict[Any, Any], bool]:
-    """
-    Unpack and decode Ultra packets using CCSDS format and XTCE packet definitions.
+    Unpack and decode Ultra TOF packets.
 
     Parameters
     ----------
-    data : list
-        Grouped data.
-    apid : int
-        The APID to process.
+    ds : xarray.Dataset
+        TOF dataset.
 
     Returns
     -------
-    decom_data : dict
-        A dictionary containing the decoded data.
+    dataset : xarray.Dataset
+        Dataset containing the decoded and decompressed data.
     """
-    # Strategy dict maps APIDs to their respective processing functions
-    strategy_dict = {
-        ULTRA_TOF.apid[0]: process_ultra_tof,
-        ULTRA_EVENTS.apid[0]: process_ultra_events,
-        ULTRA_AUX.apid[0]: process_ultra_aux,
-        ULTRA_RATES.apid[0]: process_ultra_rates,
+    scalar_keys = [key for key in ds.data_vars if key != "packetdata"]
+
+    decom_data: defaultdict[str, list[np.ndarray]] = defaultdict(list)
+    decom_data["packetdata"] = []
+    valid_epoch = []
+    width = cast(int, ULTRA_TOF.width)
+    mantissa_bit_length = cast(int, ULTRA_TOF.mantissa_bit_length)
+
+    for val, group in ds.groupby("epoch"):
+        if set(group["sid"].values) >= set(range(8)):
+            valid_epoch.append(val)
+            group.sortby("sid")
+
+            for key in scalar_keys:
+                decom_data[key].append(group[key].values)
+
+            image = []
+            for i in range(8):
+                binary = convert_to_binary_string(group["packetdata"].values[i])
+                decompressed = decompress_image(
+                    group["p00"].values[i],
+                    binary,
+                    width,
+                    mantissa_bit_length,
+                )
+                image.append(decompressed)
+
+            decom_data["packetdata"].append(np.stack(image))
+
+    for key in scalar_keys:
+        decom_data[key] = np.stack(decom_data[key])
+
+    decom_data["packetdata"] = np.stack(decom_data["packetdata"])
+
+    coords = {
+        "epoch": np.array(valid_epoch, dtype=np.uint64),
+        "sid": np.arange(8),
+        "row": xr.DataArray(np.arange(54), dims=["row"], name="row"),
+        "column": xr.DataArray(np.arange(180), dims=["column"], name="column"),
     }
 
-    sorted_packets = sort_by_time(data, "SHCOARSE")
+    dataset = xr.Dataset(coords=coords)
+
+    # Add scalar keys (2D: epoch x sid)
+    for key in scalar_keys:
+        dataset[key.upper()] = xr.DataArray(
```

Comment:
It looked like you changed most of the other places to be lowercase variable names now, this one too?
```suggestion
        dataset[key] = xr.DataArray(
```